### ISSUE TITLE
Cannot marshal nor unmarshal dates

### ISSUE THREAD

**Issue by bLuka** [apache/cassandra-gocql-driver#1148]
Hi there!

I may have found an interesting issue with dates. I hope I can help!

### What version of Cassandra are you using?

That's the 3.11.2.

### What version of Gocql are you using?

The latest version at the time (June, 17, e06f8c1)

### What did you do?

I tried to select entries containing a `date` field into a map, or inserting them into a table.

#### Selection

During a SELECT, I had a panic in reflect, called from `gocql.NativeType.New` (marshal.go:2005). After some debugging, I found `t.typ` was `custom`, `t.proto` was `3` and `t.custom` was `org.apache.cassandra.db.marshal.SimpleDateType`.

The panic was caused by `goType(t) = <nil>`.

This came from my `date` fields in my SELECT and everything was fine when I removed the date fields from the selection.

#### Insertion

In the insertion, I tried inserting values by building a `time.Time` from my code; here's the error returned by `gocql.Query.Exec()`:

`can not marshal time.Time into custom(org.apache.cassandra.db.marshal.SimpleDateType)`

### What did you expect to see?

No error at all, and my values correctly inserted or selected.

#### At selection:

A panic:

```
reflect.New(0x0, 0x0, 0x0, 0x0, 0x411ae9)
        /usr/lib/go/src/reflect/value.go:2185 +0xde
github.com/gocql/gocql.NativeType.New(0xc42007da03, 0x0, 0xc4201f6150, 0x2e, 0xc42007da01, 0xc4200e6780)
       github.com/gocql/gocql/marshal.go:2005 +0x265
```

#### At insertion:

`can not marshal time.Time into custom(org.apache.cassandra.db.marshal.SimpleDateType)`

**Comment by alourie** [comment]
@bLuka hey there

Would you mind elaborating a bit more on your use case? I just tried simple C* node with a simple table with a `date` field; simple fetching into a `time.Time` works fine.

Could you maybe show a few code snippets? It would help with debugging the issue.

Thanks!

**Comment by bLuka** [comment]
Here's a test with my current configuration.

I created a sample table in Cassandra:
```
cqlsh> create table test (id int PRIMARY KEY, date date );
cqlsh> INSERT INTO test (id, date) VALUES (0, toDate(Now()));
```

The code:
```Golang
   var iter = session.Query("SELECT * FROM test;").Iter()

   for {
      entry := make(map[string]interface{})

      if !iter.MapScan(entry) {
         break
      }
      log.Println(entry)
   }
```

The error happens at the `MapScan()` line.

**Comment by bLuka** [comment]
I just had a success trying over the database with a sample code.

I'll try to isolate the error and come here with new results.

**Comment by bLuka** [comment]
The error came from a configuration I had to set for my production cluster: `cluster.ProtoVersion = 3`

Without this configuration, I got this error when I call `cluster.CreateSession()`:
```
gocql:
unable to create session:
unable to discover protocol version:
no response to connection startup within timeout
```

If I set `cluster.ProtoVersion = 4`, no error happen.

You may close this issue, but doesn't the ProtoVersion has to be the latest available version by default?

**Comment by alourie** [comment]
@bLuka the ProtoVersion is detected automatically by using the highest one (4 atm) to connect to one of the nodes, and if that fails with `ProtocolError` it interprets the error to find one working. In your example, though, it seems it couldn't connect to the nodes at all (`no response to connection startup within timeout`), so that's why it didn't detect any protoVersion. Can you provide more details on your cluster? Why wouldn't it connect without providing the protocol?

**Comment by bLuka** [comment]
It may be due to port filtering I did on the cluster.

Only one IP (mine) on a single port (`tcp:9042`) is allowed to connect to the cluster, on only one of the three nodes, in development.

I often see some `no response to connection startup within timeout` appearing randomly at startup (which are fixed by restarting the connexion until it works).

The cluster is hosted on a Google Cloud Compute Engine instance. I use the default Google Cloud Platform's networks interface to handle the port filtering.

I fear this is weak speaking of the amount of given details; is there anything precise I could give you to help understand the issue?

**Comment by alourie** [comment]
@bLuka it is strange indeed. The moment it connects, it negotiates the protocol version. So by virtue of reporting that it has connected, it should be getting the protocol. I'll have a look through the code again tomorrow.

**Comment by martin-sucha** [comment]
Reproduced the selection panic with Cassandra 3.11.11 and gocql `v0.0.0-20211222152519-ea66763d643a`

**Comment by martin-sucha** [comment]
I don't think we need to add support for date type from protocol version 3 since all currently supported Cassandra versions support at least version 4 of CQL protocol and dates work just fine using protocol version 4. However, we should not panic in case there is no Go type available for the CQL type. I opened #1593 to address that.

**Comment by OleksiienkoMykyta** [comment]
Fixed by following PR (https://github.com/apache/cassandra-gocql-driver/pull/1593).
Should be closed.
